### ISSUE TITLE
IEDriver dismisses unhandled prompt by default

### ISSUE THREAD

**Issue by DOlllDb** [SeleniumHQ/selenium#4807]
## Meta -
OS:  Windows 7

Selenium Version:  
3.5.3, IEDriverServer 3.5.1

Browser:  
Internet Explorer

Browser Version:  
11.0.9600.18762

## Expected Behavior
I get an allert and it not closes

## Actual Behavior -
IE dismisses my alert so that my test fails when I actually try to check that my alert is preset and check it's text;  It was ok on selenium 2.53.1

## Steps to reproduce -
1. Create IE web driver like this:

```java
DesiredCapabilities capabilities = new DesiredCapabilities();
capabilities.setBrowserName("internet explorer");
capabilities.setCapability("disable-popup-blocking", true);
capabilities.setCapability(CapabilityType.VERSION, "EDGE");
capabilities.setCapability(CapabilityType.PLATFORM, "WINDOWS");
capabilities.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);
capabilities.setCapability(InternetExplorerDriver.IGNORE_ZOOM_SETTING, true);
capabilities.setCapability(InternetExplorerDriver.INTRODUCE_FLAKINESS_BY_IGNORING_SECURITY_DOMAINS, true);
capabilities.setCapability(InternetExplorerDriver.NATIVE_EVENTS, Methods.getCheckBoxState(IE_NATIVE_EVENTS) != 0);
capabilities.setCapability(InternetExplorerDriver.IE_ENSURE_CLEAN_SESSION, true);
capabilities.setCapability(InternetExplorerDriver.REQUIRE_WINDOW_FOCUS, Methods.getCheckBoxState(IE_REQUIRE_FOCUS) != 0);
capabilities.setCapability(CapabilityType.UNEXPECTED_ALERT_BEHAVIOUR, UnexpectedAlertBehaviour.IGNORE);
WebDriver driver = new InternetExplorerDriver(capabilities);
```

2. I have a page (which I cant provide here), so when I don't fill a field and press a button I get an Alert (default ie modal window). So I do click that button without filling a field to check alert text.

**Comment by luke-hill** [comment]
Could you try doing this on `the-internet` the page which should contain nearly all elements you need
`https://the-internet.herokuapp.com/`

**Comment by DOlllDb** [comment]
Sorry for confusion. I debugged a little bit and found what really happens. It doesn't really matter if it's iedriver or any other driver (checked FF and chrome).
So I press a button that triggers an event that creates an Alert - alert("some text");
Then I have a pageLoad method that checks whether I have ajax loader and if it dissapears. In that method I have driver.findElements(By.tagName(body)) and I check css value of that element (but that doesn't really matter at that point).
What I really want to do is for my alert to not close whenever a call to webdriver's findElement or FindElements happens, because otherwise I wouldn't be able to check my alerts.

P.S. Is there a way to set a parameter so that webdriver works like in a legacy (<3.0) version driver. I cant just remove my method that checks whether my page is loaded, it's not an option and I can't move to selenium 3 cuz of a problem I described. 
code to reproduce:
1. create a driver like in step one of this issue decription:
2. open uri: driver.get("http://www.gospelweb.co.uk/fowlerfamily/alerts2.html"); - it's a page with alert("some text") whenever a page is loaded (googled it because I couldn't find appropriate page on the-internet. Hope it suits you
3. call driver.findElements(By.xpath("//body")); - alert closes there
4. Assert that alert is present - can't check my alert cuz it got closed on step 3.

**Comment by p0deje** [comment]
I cannot reproduce the issue with Ruby bindings and Firefox.

**Comment by p0deje** [comment]
Can you try without this capability:

```java
capabilities.setCapability(CapabilityType.UNEXPECTED_ALERT_BEHAVIOUR, UnexpectedAlertBehaviour.IGNORE);
```

**Comment by DOlllDb** [comment]
Checked IE, Chrome and FF just now. Even upgraded to 3.6.0. and downwloaded IEDriverServer 3.6.0.
True, that I can't reproduce it neither on chrome nor FF. But still there is an issue in IE. Alert just dissapears:
even tried without any capabilities:
WebDriver drv = new InternetExplorerDriver();
driver.get("http://www.gospelweb.co.uk/fowlerfamily/alerts2.html");
// Alert is present atm
driver.findElement(By.xpath("body"));
// Alert dissapears
driver.switchTo().alert();
// you get an exception NoSuchAlertPresent;

**Comment by barancev** [comment]
I can't reproduce it with IE 11 and IEDriverServer 3.6.0.0
![image](https://user-images.githubusercontent.com/617090/31283016-06a33de0-aabd-11e7-8269-dab7876ce7df.png)

**Comment by DOlllDb** [comment]
Tried testing this at home, worked perfectly. But I still get the same result as before. Modal dialog closes. Also tried to reset explorer to default settings, to no avail. Also asked my co-worker to test this issue on his PC, got same results (alert closed automatically).
So I ran this test as you suggested with debug logging level
[iedriver.log](https://github.com/SeleniumHQ/selenium/files/1367971/iedriver.log)
P.S. As you can see, I'm opening a page with alert from alert2.mht (also tried alert2.html) because I don't have internet access on my main PC at work so I had to download the page.

**Comment by DOlllDb** [comment]
Any progress reproducing my issue?

**Comment by p0deje** [comment]
I now understand what the problem is. You get an unexpected alert exception but *also* alert is automatically dismissed. This does not comply with the specification which says:

> By default user prompts are not handled automatically unless a user prompt handler has been defined. When a user prompt appears, it is the task of the subsequent command to handle it. If the subsequent requested command is not one listed in this chapter, an unexpected alert open error will be returned.

I've attempted to fix this in c774792d543b but that's the first time I touch IEDriver so I'll need a help from @jimevans to get this merged.

In the meantime, you can work around this behavior by using "ignore" unhandled prompt behavior. I tested with Ruby bindings, but in Java it should look like this:

```java
DesiredCapabilities capabilities = new DesiredCapabilities();
capabilities.setCapability(CapabilityType.UNHANDLED_PROMPT_BEHAVIOUR, UnexpectedAlertBehaviour.IGNORE);
WebDriver driver = new InternetExplorerDriver(capabilities);
```

Note that `CapabilityType.UNEXPECTED_ALERT_BEHAVIOR` is not respected by IEDriver. Such capability does not exist in WebDriver specification.

**Comment by p0deje** [comment]
After further investigation, it turned out the spec is a bit misleading and the default behavior is correctly implemented by IEDriver. I've filled a bug to fix the spec in w3c/webdriver#1135.

Meanwhile, closing this issue as you need to use `CapabilityType.UNHANDLED_PROMPT_BEHAVIOUR` instead of `CapabilityType.UNEXPECTED_ALERT_BEHAVIOUR`. Everything else works as expected.
